DMApp.Keystone/UserExecutable.cs

```diff
@@ -0,0 +1,141 @@
+            fs.Directory.CopyRecursive(pathToUserExecutableDir, userProgramPathForPackaging);
+
+            // Defaulting data:
+            if (String.IsNullOrWhiteSpace(toolMetaData.ToolName)) toolMetaData.ToolName = $"Skyline.DataMiner.Keystone.{programName}";
+            if (String.IsNullOrWhiteSpace(toolMetaData.ToolCommand))
+            {
+                string cleanedProgramName = Regex.Replace(programName, "[^a-zA-Z0-9 ]", "");
+                string cleanedCommandName = Regex.Replace(cleanedProgramName, @"\s+", "-").ToLower();
+                toolMetaData.ToolCommand = $"dataminer-keystone-{cleanedCommandName}";
+            }
+
+            if (String.IsNullOrWhiteSpace(toolMetaData.ToolVersion))
+            {
+                // Handles IncludeSourceRevisionInInformationalVersion scenarios.
+                var exeVersion = fs.File.GetFileProductVersion(programPath).Split('+')[0];
+                if (String.IsNullOrWhiteSpace(exeVersion))
+                {
+                    exeVersion = fs.File.GetFileVersion(programPath);
+                }
+
+                if (String.IsNullOrWhiteSpace(exeVersion))
+                {
+                    exeVersion = "1.0.0";
+                }
+
+                toolMetaData.ToolVersion = exeVersion;
+            }
+
+            if (String.IsNullOrWhiteSpace(toolMetaData.Company)) toolMetaData.Company = "Undefined";
+            if (String.IsNullOrWhiteSpace(toolMetaData.Authors)) toolMetaData.Authors = "Undefined";
+
+            try
+            {
+                Dictionary<string, string> placeholdersToReplace = new()
+                {
+                    { nameof(ToolMetaData.ToolName), toolMetaData.ToolName },
+                    { nameof(ToolMetaData.ToolCommand), toolMetaData.ToolCommand },
+                    { nameof(ToolMetaData.Authors), toolMetaData.Authors },
+                    { nameof(ToolMetaData.Company), toolMetaData.Company },
+                    { nameof(ToolMetaData.ToolVersion), toolMetaData.ToolVersion },
+                    { "ProgramName", programName },
+                    { "ProgramNameShimmy", userProgramFolderName }
+                };
+
+                foreach (var topLevelFile in fs.Directory.EnumerateFiles(shimmyPathForPackaging))
+                {
+                    var extension = fs.Path.GetExtension(topLevelFile);
+                    if (extension != ".md" && extension != ".cs" && extension != ".txt" && extension != ".csproj") continue;
+
+                    string fileContent = fs.File.ReadAllText(topLevelFile);
+                    fileContent = ReplaceAllVariables(fileContent, placeholdersToReplace);
+                    fs.File.WriteAllText(topLevelFile, fileContent);
+                }
+
+                var packResult = dotnet.Run($"pack \"{shimmyPathForPackaging}/ExeShim.csproj\" --output \"{toolMetaData.OutputDirectory}\"");
```

Comment:
When the program name contains spaces, this eventually returns an error:
![image](https://github.com/SkylineCommunications/Skyline.DataMiner.CICD.Packages/assets/97458010/bb10907a-3d1a-46c0-8291-e3f2b9376730)

Maybe some early validation couldn't hurt as people tend to use spaces a lot? Then it will also be a clearer error message.

Reply:
I've sanitized the inputs instead of throwing errors for ease of use.